File: 문제풀이/chap6-21_25/prob6-23.c
#include <stdio.h>
#include <stdbool.h>

long long L, R; bool isprime[500009];

int main() {
	// 입력
	scanf("%lld%lld", &L, &R);

	// 배열 초기화
	long long i, j;
	for (i = 0; i <= R - L; i++) {
		isprime[i] = true;
	}

	// L = 1  때의 조건 분기(코너 케이스)
	if (L == 1) {
		isprime[0] = false;
	}

	// 에라토스테네스의 체
	for (i = 2; i * i <= R; i++) {
		long long min_value = ((L + i - 1) / i) * i; // L 이상에서 최소인 i의 배수
		// L 이상 R 이하에서 i를 제외한 i의 배수에 X 표시
		for (j = min_value; j <= R; j += i) {
			if (j == i) continue;
			isprime[j - L] = false;
		}
	}

	// 갯수 세서 출력하기
	long long answer = 0;
	for (i = 0; i <= R - L; i++) {
		if (isprime[i] == true) {
			answer += 1;
		}
	}
	printf("%lld\n", answer);
	return 0;
}

File: 문제풀이/chap6-21_25/prob6-23.py
# 입력
L, R = map(int, input().split())

# 배열 초기화/L=1  때의 조건 분기
isprime = [ True ] * (R - L + 1)
if L == 1:
	isprime[0] = False

# 에라토스테네스의 체
LIMIT = int(R ** 0.5)
for i in range(2, LIMIT + 1):
	min_value = ((L + i - 1) // i) * i
	for j in range(min_value, R + 1, i):
		if j == i:
			continue
		isprime[j - L] = False

# 갯수 세서 출력하기
answer = 0
for i in range(R - L + 1):
	if isprime[i] == True:
		answer += 1
print(answer)